File: c-interface/src/Input.cpp
#include "Astral.Canvas/Input.h"
#include "Input/Input.hpp"
#include "GLFW/glfw3.h"

exportC bool AstralCanvasInput_IsKeyDown(AstralCanvas_Keys key)
{
    return AstralCanvas::Input_IsKeyDown((AstralCanvas::Keys)key);
}

exportC bool AstralCanvasInput_IsKeyPressed(AstralCanvas_Keys key)
{
    return AstralCanvas::Input_IsKeyPressed((AstralCanvas::Keys)key);
}

exportC bool AstralCanvasInput_IsKeyReleased(AstralCanvas_Keys key)
{
    return AstralCanvas::Input_IsKeyReleased((AstralCanvas::Keys)key);
}

exportC bool AstralCanvasInput_IsMouseDown(AstralCanvas_MouseButtons button)
{
    return AstralCanvas::Input_IsMouseDown((AstralCanvas::MouseButtons)button);
}

exportC bool AstralCanvasInput_IsMousePressed(AstralCanvas_MouseButtons button)
{
    return AstralCanvas::Input_IsMousePressed((AstralCanvas::MouseButtons)button);
}

exportC bool AstralCanvasInput_IsMouseReleased(AstralCanvas_MouseButtons button)
{
    return AstralCanvas::Input_IsMouseReleased((AstralCanvas::MouseButtons)button);
}

exportC void AstralCanvasInput_SimulateMousePress(AstralCanvas_MouseButtons button)
{
    AstralCanvas::Input_SimulateMousePress((AstralCanvas::MouseButtons)button);
}

exportC void AstralCanvasInput_SimulateMouseRelease(AstralCanvas_MouseButtons button)
{
    AstralCanvas::Input_SimulateMouseRelease((AstralCanvas::MouseButtons)button);
}

exportC bool AstralCanvasInput_ControllerIsButtonDown(const u32 controllerIndex, const AstralCanvas_ControllerButtons button)
{
    return AstralCanvas::Input_ControllerIsButtonDown(controllerIndex, (AstralCanvas::ControllerButtons)button);
}
exportC bool AstralCanvasInput_ControllerIsButtonPress(const u32 controllerIndex, const AstralCanvas_ControllerButtons button)
{
    return AstralCanvas::Input_ControllerIsButtonPress(controllerIndex, (AstralCanvas::ControllerButtons)button);
}
exportC bool AstralCanvasInput_ControllerIsButtonRelease(const u32 controllerIndex, const AstralCanvas_ControllerButtons button)
{
    return AstralCanvas::Input_ControllerIsButtonRelease(controllerIndex, (AstralCanvas::ControllerButtons)button);
}
exportC bool AstralCanvasInput_ControllerIsR2Down(const u32 controllerIndex)
{
    return AstralCanvas::Input_ControllerIsR2Down(controllerIndex);
}
exportC float AstralCanvasInput_ControllerGetR2DownAmount(const u32 controllerIndex)
{
    return AstralCanvas::Input_ControllerGetR2DownAmount(controllerIndex);
}
exportC bool AstralCanvasInput_ControllerIsL2Down(const u32 controllerIndex)
{
    return AstralCanvas::Input_ControllerIsL2Down(controllerIndex);
}
exportC float AstralCanvasInput_ControllerGetL2DownAmount(const u32 controllerIndex)
{
    return AstralCanvas::Input_ControllerGetL2DownAmount(controllerIndex);
}
exportC bool AstralCanvasInput_ControllerIsConnected(const u32 controllerIndex)
{
    return AstralCanvas::Input_ControllerIsConnected(controllerIndex);
}
exportC AstralCanvasVector2 AstralCanvasInput_GetLeftStickAxis(const u32 controllerIndex)
{
    Maths::Vec2 result = AstralCanvas::Input_GetLeftStickAxis(controllerIndex);
    return {result.X, result.Y};
}
exportC AstralCanvasVector2 AstralCanvasInput_GetRightStickAxis(const u32 controllerIndex)
{
    Maths::Vec2 result = AstralCanvas::Input_GetRightStickAxis(controllerIndex);
    return {result.X, result.Y};
}

exportC bool AstralCanvasInput_ControllerIsL2Pressed(const u32 controllerIndex)
{
    return AstralCanvas::Input_ControllerIsL2Pressed(controllerIndex);
}
exportC bool AstralCanvasInput_ControllerIsR2Pressed(const u32 controllerIndex)
{
    return AstralCanvas::Input_ControllerIsR2Pressed(controllerIndex);
}
exportC bool AstralCanvasInput_ControllerIsL2Released(const u32 controllerIndex)
{
    return AstralCanvas::Input_ControllerIsL2Released(controllerIndex);
}
exportC bool AstralCanvasInput_ControllerIsR2Released(const u32 controllerIndex)
{
    return AstralCanvas::Input_ControllerIsR2Released(controllerIndex);
}

exportC AstralCanvasVector2 AstralCanvasInput_GetMousePosition()
{
    Maths::Vec2 result = AstralCanvas::Input_GetWindow()->windowInputState.mousePosition;

    return {result.X, result.Y};
}
exportC AstralCanvasVector2 AstralCanvasInput_GetMouseScroll()
{
    Maths::Vec2 result = AstralCanvas::Input_GetWindow()->windowInputState.scroll;
    return {result.X, result.Y};
}
exportC void AstralCanvasInput_SetMousePosition(AstralCanvasVector2 position)
{
    glfwSetCursorPos((GLFWwindow*)AstralCanvas::Input_GetWindow()->handle, (double)position.X, (double)position.Y);
}
exportC u32 AstralCanvasInput_AwaitInputChar()
{
    return AstralCanvas::Input_GetWindow()->windowInputState.textInputCharacter;
}

File: bindings/csharp/Input.cs
using System;
using System.Runtime.CompilerServices;
using System.Numerics;

namespace Astral.Canvas
{
    public static class Input
    {

        /// <summary>
        /// Returns true if a key is just pressed or is being held down.
        /// </summary>
        /// <param name="key">The key to check</param>
        /// <returns></returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool IsKeyDown(Keys key) => AstralCanvas.Input_IsKeyDown(key);

        /// <summary>
        /// Returns true for a frame if a key has just been pressed
        /// </summary>
        /// <param name="key">The key to check</param>
        /// <returns></returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool IsKeyPressed(Keys key) => AstralCanvas.Input_IsKeyPressed(key);

        /// <summary>
        /// Returns true for a frame if a key has just been released
        /// </summary>
        /// <param name="key">The key to check</param>
        /// <returns></returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool IsKeyReleased(Keys key) => AstralCanvas.Input_IsKeyReleased(key);

        /// <summary>
        /// Returns true if the specified mouse button is held down.
        /// </summary>
        /// <param name="button">The mouse button to check</param>
        /// <returns></returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool IsMouseDown(MouseButtons button) => AstralCanvas.Input_IsMouseDown(button);

        /// <summary>
        /// Returns true for a frame if the specified mouse button has just been pressed
        /// </summary>
        /// <param name="button"></param>
        /// <returns></returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool IsMousePressed(MouseButtons button) => AstralCanvas.Input_IsMousePressed(button);

        /// <summary>
        /// Returns true for a frame if the specified mouse button has just been released
        /// </summary>
        /// <param name="button"></param>
        /// <returns></returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool IsMouseReleased(MouseButtons button) => AstralCanvas.Input_IsMouseReleased(button);
        /// <summary>
        /// Returns a character whenever the user types an input/holds down the button
        /// </summary>
        /// <returns></returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static uint AwaitInputChar() => AstralCanvas.Input_AwaitInputChar();

        public static Vector2 mousePosition => AstralCanvas.Input_GetMousePosition();
        public static Vector2 worldMousePosition;

        /// <summary>
        /// Simulates the user pressing the mouse, thus IsMouseDown will become 
        /// true and IsMousePressed as well, for that frame. 
        /// <br>Up to the user to call SimulateMouseRelease to release the mouse.</br>
        /// </summary>
        /// <param name="button"></param>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static void SimulateMousePress(MouseButtons button) => AstralCanvas.Input_SimulateMousePress(button);
        /// <summary>
        /// Simulates the user releasing the mouse, thus IsMouseDown will become 
        /// false and IsMouseReleased as well, for that frame. 
        /// <br>Up to the user to call SimulateMouseRelease to release the mouse.</br>
        /// </summary>
        /// <param name="button"></param>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static void SimulateMouseRelease(MouseButtons button) => AstralCanvas.Input_SimulateMouseRelease(button);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static void SetMousePosition(Vector2 position) => AstralCanvas.Input_SetMousePosition(position);
        public static Vector2 mouseScroll => AstralCanvas.Input_GetMouseScroll();

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsConnected(uint controllerID) => AstralCanvas.Input_ControllerIsConnected(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector2 ControllerGetLeftStickAxis(uint controllerID) => AstralCanvas.Input_GetLeftStickAxis(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector2 ControllerGetRightStickAxis(uint controllerID) => AstralCanvas.Input_GetRightStickAxis(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float ControllerGetL2DownAmount(uint controllerID) => AstralCanvas.Input_ControllerGetL2DownAmount(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float ControllerGetR2DownAmount(uint controllerID) => AstralCanvas.Input_ControllerGetR2DownAmount(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsL2Down(uint controllerID) => AstralCanvas.Input_ControllerIsL2Down(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsR2Down(uint controllerID) => AstralCanvas.Input_ControllerIsR2Down(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsL2Pressed(uint controllerID) => AstralCanvas.Input_ControllerIsL2Pressed(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsR2Pressed(uint controllerID) => AstralCanvas.Input_ControllerIsR2Pressed(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsL2Released(uint controllerID) => AstralCanvas.Input_ControllerIsL2Released(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsR2Released(uint controllerID) => AstralCanvas.Input_ControllerIsR2Released(controllerID);
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsButtonDown(uint controllerID, ControllerButtons button)
        {
            return AstralCanvas.Input_ControllerIsButtonDown(controllerID, button);
        }
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsButtonPressed(uint controllerID, ControllerButtons button)
        {
            return AstralCanvas.Input_ControllerIsButtonPress(controllerID, button);
        }
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool ControllerIsButtonReleased(uint controllerID, ControllerButtons button)
        {
            return AstralCanvas.Input_ControllerIsButtonRelease(controllerID, button);
        }
    }
}
